python/tvm/relay/tensorrt.py

```diff
@@ -0,0 +1,194 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name,arguments-differ,no-else-return,unused-argument,missing-docstring
+"""
+Relay TensorRT codegen.
+"""
+import tvm
+from tvm import relay
+from tvm.relay.expr import Call, Constant
+
+from . import _transform
+from .expr_functor import ExprMutator
+
+def _bind_params(func, params):
+    """
+    Bind the params to the expression as constants.
+    """
+    name_dict = {}
+    for arg in func.params:
+        name = arg.name_hint
+        if name in name_dict:
+            name_dict[name] = None
+        else:
+            name_dict[name] = arg
+    bind_dict = {}
+    for k, v in params.items():
+        if k not in name_dict:
+            continue
+        arg = name_dict[k]
+        if arg is None:
+            raise ValueError("Multiple args in the function have name %s" % k)
+        bind_dict[arg] = relay.expr.const(v)
+    return relay.expr.bind(func, bind_dict)
+
+class LegalizeLayoutTranform(ExprMutator):
+    """
+    Legalize Relay layout transforms to transpose ops to simplify TensorRT conversion.
```

Comment:
Do we still need to use Legalize and `relay.transpose`? Can we use layout transform to convert source Relay graph to what TensorRT expects?

Reply:
I think its better to leverage relay's pass ability to convert `layout_transform` op to the more standard `transpose` ops. This way we only need to write one TrtOpConverter for `transpose`. If we didn't perform this legalize, I would need to write an additional TrtOpConverter for `layout_transform` which would be nearly identical to the one for `transpose`.

This feature of relay is very useful. For example, TRT recently announced that they won't support INT8 for matmul/fully connected layer and to just use 1x1 Conv instead (https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#optimize-layer). So in the future, I plan to have a similar pass to convert all matmul/dense layers into convolutions to take advantage of this. At that point I won't need a converter for dense anymore since everything would go to conv.